LDAP Group Mapping

darolfes opened the pull request:
Assume the following LDAP configuration:
```ini
org.opencastproject.userdirectory.ldap.roleattributes=memberOf
org.opencastproject.userdirectory.ldap.roleprefix=ROLE_
```
and the following LDAP User:
```ini
memberOf=student, staff, whatever
```

This pull request implements 3 features related to LDAP.

1. Currently, Opencast would try to add all `memberOf` attributes as a group, this leads to warnings if the group doesn't exist.
A new configuration key is added, which checks whether a attribute has a specific prefix, if the prefix is present the attribute is added as a group, otherwise it is only added as a role.
Default config:
```ini
org.opencastproject.userdirectory.ldap.groupprefix=ROLE_GROUP_
```

2. A Map from LDAP attributes to roles, e.g.:
```ini
org.opencastproject.userdirectory.ldap.roles.staff=ROLE_ADMIN
```
This would add the `ROLE_ADMIN`  to the user, because the user has the `staff` attribute.

3. An additional Map, but for groups.

darolfes committed:
LDAP Group Mapping

lkiesow reviewed (changes requested):
The new group prefix works great and is a huge help in avoiding a log full of spam when using LDAP.

But can you slightly change the way this works? Right now, you have to configure `org.opencastproject.userdirectory.ldap.roleattributes=abc, xyz` which is then used as the basis for the role and group mapping. This means that whenever you want to map something, the attributes itself are always added as roles and it also makes it really hard (or impossible) to map multiple attributes differently.

```properties
org.opencastproject.userdirectory.ldap.map.identifier1.attribute=memberOf
org.opencastproject.userdirectory.ldap.map.identifier1.value=staff
org.opencastproject.userdirectory.ldap.map.identifier1.roles=ROLE_EXAMPLE,ROLE_EXAMPLE2
org.opencastproject.userdirectory.ldap.map.identifier1.groups=ROLE_GROUP_EXAMPLE

org.opencastproject.userdirectory.ldap.map.student.attribute=memberOf
org.opencastproject.userdirectory.ldap.map.student.value=student
org.opencastproject.userdirectory.ldap.map.student.roles=ROLE_STUDENT
org.opencastproject.userdirectory.ldap.map.student.groups=ROLE_GROUP_STUDENT
```

gregorydlogan commented:
Reminder: As a feature this has until 2020-10-05 to be merged for it to make it in 9.0!

lkiesow commented:
@darolfes, any chance to get this done until Monday?

lkiesow commented:
@darolfes, unfortunately, I can't get a user to get any roles anymore.

The configuration I tries is
https://data.lkiesow.io/opencast/ldapconfig/ldapconfig-pr-1634-2020-10-05/

The `org.opencastproject.userdirectory.ldap.extra.roles=ROLE_ANONYMOUS` alone should ensure that `jmacy` would at least get the role `ROLE_ANONYMOUS`, shouldn't it?

sternenseemann commented:
Any news on this? Merging this would save us a lot of trouble with LDAP groups :)

KatrinIhler commented:
@darolfes @lkiesow Is this still being worked on?

darolfes commented:
Yes, I can probably push a update next week.

darolfes committed:
LDAP Group Mapping: workaround for the missing role bug

darolfes commented:
I updated the pull request.

A slight change in the order of lines in the configuration resulted in the roles being dropped/ignored after `LdapUserProviderInstance.java` finished.
This strange behavior seems like a bug in Opencast, but I'm not completely sure.
As a workaround I moved all the role assigning code into the `OpencastLdapAuthoritiesPopulator`, which behaves as expected.
This also seems to be the way spring-security expects role assignment to be handled anyway.

Arnei reviewed (approved):
Tested this in a local AllInOne installation with the LDAP server from forumsys and all settings were working as expected.

Code looks fine too.

lkiesow commented:
> Tested this in a local AllInOne installation with the LDAP server from forumsys and all settings were working as expected.

Then, let's get this merged.
